Update phpcpd to ~6.0.3 to be PHP8 compatible

hws47a opened the pull request:
### Description (*)
<!---
    Please provide a description of the changes proposed in the pull request.
    Letting us know what has changed and why it needed changing will help us validate this pull request.
-->
Update phpcpd to ~6.0.3 with minimal changes to other dependencies including update phpunit/php-timer from ^3.0 to ^5.0 and phpunit from 9.1 to 9.2

Changes included:

```
  - Updating phpunit/php-timer (3.1.4 => 5.0.3)
  - Installing sebastian/cli-parser (1.0.1)
  - Updating sebastian/phpcpd (5.0.2 => 6.0.3)
  - Updating phpunit/phpunit (9.1.5 => 9.2.6)
```

### Related Pull Requests
https://github.com/magento/partners-magento2ee/pull/501
https://github.com/magento/partners-magento2b2b/pull/563
<!-- related pull request placeholder -->

### Fixed Issues (if relevant)
<!---
    If relevant, please provide a list of fixed issues in the format magento/magento2#<issue_number>.
    There could be 1 or more issues linked here and it will help us find some more information about the reasoning behind this change.
-->
1. Fixes magento/magento2#31344

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

hws47a commented:
@magento run all tests

hws47a commented:
@sivaschenko Static Tests passed 👏
and phpcpd is not used anywhere else

hws47a commented:
@sivaschenko some integration test has failed for Integration B2B but I cannot find it in partners-b2b repository and php8-develop branch is not available there.

Could you help me with this?

hws47a commented:
@magento run Integration Tests

ihor-sviziev commented:
@magento run Integration Tests

hws47a commented:
@ihor-sviziev it seems that all passed

ihor-sviziev commented:
@hws47a, could you explain why we need to update PHPUnit in the scope of this PR?

hws47a commented:
@ihor-sviziev 

You may see that in composer.json only phpcpd was changed and PHPUnit was allowed to be any ^9 version.
phpcpd was not updating without phpunit as they had different dependencies for phpunit/php-timer.
I found that PHPUnit 9.2 is the first version which allows required phpunit/php-timer and updated to it.

So a bit newer PHPUnit is a requirement for phpcpd 6.0.3

sivaschenko commented:
@hws47a thanks for the update, can you please also update the composer.json and lock in EE repo and link the EE PR as related

sivaschenko commented:
@hws47a btw, the php8-develop should be present for all repositories, i.e here is B2B branch https://github.com/magento/partners-magento2b2b/tree/php8-develop

hws47a commented:
> @hws47a thanks for the update, can you please also update the composer.json and lock in EE repo and link the EE PR as related

Done

hws47a commented:
@magento run Static Tests

hws47a commented:
@magento run Functional Tests B2B, Functional Tests EE, Static Tests

ihor-sviziev commented:
@magento run Functional Tests B2B, Static Tests

ihor-sviziev commented:
@hws47a, static tests are still failing. Could you check, please?

hws47a commented:
@magento run all tests

hws47a commented:
@magento run all tests

hws47a commented:
@magento run Database Compare, Functional Tests B2B, Functional Tests EE, Integration Tests, Magento Health Index, Static Tests, Unit Tests, WebAPI Tests

ihor-sviziev commented:
@sivaschenko any thoughts why tests are failing with no build results?

sivaschenko commented:
@hws47a @ihor-sviziev tests have failed due to conflict in the EE PR

hws47a commented:
@magento run all tests

ihor-sviziev commented:
@magento run all tests

ihor-sviziev reviewed (changes requested):
Could you resolve conflicts in EE repo and run all tests again?

ihor-sviziev commented:
@magento run Functional Tests B2B, Functional Tests EE, Static Tests

ihor-sviziev commented:
@magento run Functional Tests B2B, Functional Tests EE, Static Tests

ihor-sviziev reviewed (changes requested):
Hi @hws47a,
It seems like static tests are failing due to some changes in the phpcpd, or maybe in some other dependency that you updated. Could you debug the test failure and fix it?

hws47a commented:
@magento run all tests

hws47a commented:
@ihor-sviziev I don't know how exactly tests are run for B2B. It's weird that they pass on CE and EE but break on B2B. I would need help of @sivaschenko to either find the issue either help me reproduce the issue locally.

ihor-sviziev commented:
@sivaschenko could you help with fixing static test failures?

sivaschenko commented:
@magento run all tests

sivaschenko commented:
Switched PR base branch to 2.4-develop (all changes from php8-develop branch have been merged to 2.4-develop)

sivaschenko commented:
@hws47a @ihor-sviziev the `--names-exclude` phpcpd option was removed in version 6.0.0.

It is used here:
https://github.com/magento/magento2/blob/c46d9bfea43f6ce8fe572b4498e858f15b3319cb/dev/tests/static/framework/Magento/TestFramework/CodingStandard/Tool/CopyPasteDetector.php#L87

Here are the release notes:
https://github.com/sebastianbergmann/phpcpd/blob/master/ChangeLog.md#600---2020-08-13

So we have to switch to the `--exclude` option instead.

The issue with using only the `--exclude` option is that it processes only the absolute paths (however including glob patterns).

The paths passed to `--exclude` option are processed this way:
https://github.com/sebastianbergmann/php-file-iterator/blob/97eb187efc3560da69c5b501235cd3eb1ebfec86/src/Factory.php#L77

That is not compatible with the format of our current blacklist:
https://github.com/magento/magento2/blob/2.4-develop/dev/tests/static/testsuite/Magento/Test/Php/_files/phpcpd/blacklist/common.txt

sivaschenko commented:
@magento run Unit Tests

hws47a commented:
Working on a fix. Thank you @sivaschenko

hws47a commented:
@magento run all tests

hws47a commented:
@magento run all tests

hws47a commented:
@magento run Static Tests

hws47a commented:
@magento run Static Tests

hws47a commented:
@magento run Static Tests

hws47a commented:
@magento run all tests

hws47a commented:
Hi @sivaschenko, I've fixed phpcpd locally including ee and b2b, however it still failing in the pipeline. I've added output of the command and you may see in the pipeline that there is no output at all. So it seems like the command doesn't run a all. Could you please check it?

ihor-sviziev commented:
Let's try to re-run the tests
@magento run all tests

sivaschenko commented:
@magento run Static Tests

sivaschenko commented:
Executing PHPCPD in the phpunit process for debugging purpose

@magento run Static Tests

sivaschenko commented:
@hws47a looks like phpcpd fails to compare bitmap files also it does find a number of copy-pastes, mostly in Inventory modules

sivaschenko commented:
@magento run Static Tests

sivaschenko commented:
@magento run Static Tests

sivaschenko commented:
@magento run Static Tests

ihor-sviziev commented on dev/tests/static/framework/Magento/TestFramework/CodingStandard/Tool/CopyPasteDetector.php:
```diff
@@ -64,35 +74,103 @@ public function canRun()
+        (new PMD($this->reportFile))->processClones($clones);
+        (new Text)->printResult($clones, false);
+
+        return count($clones) == 0;
```
Plus as use strict comparison here

ihor-sviziev commented on dev/tests/static/framework/Magento/TestFramework/CodingStandard/Tool/CopyPasteDetector.php:
```diff
@@ -64,35 +74,103 @@ public function canRun()
+            $exclude[] = $files;
+        }
+
+        return array_unique(array_merge(...$exclude));
```
Technically, there might be empty array in the $exclude, so it's better to replace it with `return array_unique(array_merge([], ...$exclude));`

hws47a replied:
I've tried in interactive shell and it seems that `array_merge(...[])` doesn't throw any warnings and just returns an empty array. Do you see some PHP error while testing?

ihor-sviziev reviewed (changes requested):
Please review my comments

hws47a committed:
#31344 Update phpcpd to ^6.0.3 to be PHP8 compatible + change usage

hws47a commented:
Changed implementation of `canRun` which now checks if required packages are installed.
Also squashed all commits into one as requested by @sivaschenko

hws47a commented:
@magento run all tests

sivaschenko committed:
magento/magento2#31344: Updated phpcpd blacklist

sivaschenko commented:
@magento run all tests
